server/app/views/components/TextFormatter.java

```diff
@@ -1,190 +1,121 @@
-   * @param urlOpenAction Enum indicating whether the url should open in a new tab.
-   * @param addRequiredIndicator Whether a required indicator, in the form of a red asterisk, should
-   *     be added to the html element.
-   */
-  public static ImmutableList<DomContent> createLinksAndEscapeText(
-      String content, UrlOpenAction urlOpenAction, boolean addRequiredIndicator) {
-    // JAVASCRIPT option avoids including surrounding quotes or brackets in the URL.
-    List<Url> urls = new UrlDetector(content, UrlDetectorOptions.JAVASCRIPT).detect();
-
-    ImmutableList.Builder<DomContent> contentBuilder = ImmutableList.builder();
-    for (int i = 0; i < urls.size(); i++) {
-      Url url = urls.get(i);
-      try {
-        // While technically they could be part of the URL, trailing punctuation
-        // is more likely to be part of the surrounding text, so we strip.
-        url = Url.create(StringUtils.stripEnd(url.getOriginalUrl(), ".?!,:;"));
-      } catch (MalformedURLException e) {
-        logger.error(
-            String.format(
-                "Failed to parse URL %s after stripping trailing punctuation",
-                url.getOriginalUrl()));
-      }
-
-      int urlStartIndex = content.indexOf(url.getOriginalUrl());
-      // Find where this URL is in the text.
-      if (urlStartIndex == -1) {
-        logger.error(
-            String.format(
-                "Detected URL %s not present in actual content, %s.",
-                url.getOriginalUrl(), content));
-        continue;
-      }
+  private static final CiviFormMarkdown CIVIFORM_MARKDOWN = new CiviFormMarkdown();
 
-      // If this URL looks like it's actually an email address, skip it.
-      if (EmailValidator.getInstance().isValid(url.getOriginalUrl())) {
-        continue;
-      }
-      // If immediately following this URL there's an '@' and another URL, they could
-      // be two parts of an email address, so skip them both in case.
-      // [0] [1] [2]
-      // len: 3
-      if (i < (urls.size() - 1)) {
-        int nextUrlStartIndex = content.indexOf(urls.get(i + 1).getOriginalUrl());
-        if (content.charAt(nextUrlStartIndex - 1) == '@') {
-          i = i + 1;
-          continue;
-        }
-      }
+  /** Passes provided text through Markdown formatter. */
+  public static ImmutableList<DomContent> formatText(
+      String text, boolean preserveEmptyLines, boolean addRequiredIndicator) {
 
-      if (urlStartIndex > 0) {
-        // If it's not at the beginning, add the text from before the URL.
-        contentBuilder.add(text(content.substring(0, urlStartIndex)));
-      }
-      // Add the URL.
-      var urlTag =
-          a().withText(url.getOriginalUrl())
-              .withHref(url.getFullUrl())
-              .withClasses(
-                  "text-blue-900", "font-bold", "opacity-75", "underline", "hover:opacity-100");
-
-      if (urlOpenAction == UrlOpenAction.NewTab) {
-        urlTag
-            .withTarget("_blank")
```

Comment:
(nonblocking) Maybe I'm just missing it. How does the new parser know when to add `target="_blank"`?

Reply:
Currently, it's [always adding it ](https://github.com/civiform/civiform/pull/5644/files#diff-603ea7dbf066a2af47b7d00b6c0af09e817f85787870903c714dcbca60dc2dbbR44). I made that call because we aren't using `SameTab` anywhere right now, and I think design wise there's a preference for opening in a new tab, but I could be wrong about that. Do you think I should add the ability to specify same vs new tab back in or hold off until the need comes up?

Reply:
Having external links open in a new tab is perfect. I was just wondering what the mechanism is that sets it. The link didn't take me directly to a spot, not sure where it was pointing.

Reply:
Oh weird! I was trying to point you to line 34 (44 now) in `CiviformMarkdown.java` where we have `attributes.put("target", "_blank");`

Reply:
Oh I must have missed it. Thanks!